source/common/upstream/load_balancer_impl.cc

```diff
@@ -705,17 +705,41 @@ const HostVector& ZoneAwareLoadBalancerBase::hostSourceToHosts(HostsSource hosts
 EdfLoadBalancerBase::EdfLoadBalancerBase(
     const PrioritySet& priority_set, const PrioritySet* local_priority_set, ClusterStats& stats,
     Runtime::Loader& runtime, Random::RandomGenerator& random,
-    const envoy::config::cluster::v3::Cluster::CommonLbConfig& common_config)
+    const envoy::config::cluster::v3::Cluster::CommonLbConfig& common_config,
+    TimeSource& time_source)
     : ZoneAwareLoadBalancerBase(priority_set, local_priority_set, stats, runtime, random,
                                 common_config),
-      seed_(random_.random()) {
+      seed_(random_.random()),
+      endpoint_warming_policy_(common_config.has_slow_start_config()
+                                   ? common_config.slow_start_config().endpoint_warming_policy()
+                                   : envoy::config::cluster::v3::Cluster::CommonLbConfig::NO_WAIT),
+      slow_start_window_(std::chrono::milliseconds(
+                             common_config.has_slow_start_config()
+                                 ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(
+                                       common_config.slow_start_config(), slow_start_window, 0)
+                                 : 0) *
+                         1000),
+      time_bias_runtime_(
+          common_config.has_slow_start_config() && common_config.slow_start_config().has_time_bias()
+              ? std::make_unique<Runtime::Double>(common_config.slow_start_config().time_bias(),
+                                                  runtime)
+              : nullptr),
+      time_source_(time_source),
+      hosts_in_slow_start_(
+          std::make_shared<absl::btree_set<HostSharedPtr, orderByCreateDateDesc>>()),
+      slow_start_enabled_(slow_start_window_ > std::chrono::milliseconds(0)) {
   // We fully recompute the schedulers for a given host set here on membership change, which is
   // consistent with what other LB implementations do (e.g. thread aware).
   // The downside of a full recompute is that time complexity is O(n * log n),
   // so we will need to do better at delta tracking to scale (see
   // https://github.com/envoyproxy/envoy/issues/2874).
   priority_set.addPriorityUpdateCb(
       [this](uint32_t priority, const HostVector&, const HostVector&) { refresh(priority); });
+  priority_set.addMemberUpdateCb([this](const HostVector& hosts_added, const HostVector&) -> void {
+    if (slow_start_enabled_) {
+      recalculateHostsInSlowStart(hosts_added);
```

Comment:
hosts in slow start are now recalculated in membership update callback + in refresh upon priority update (to catch health state change)